ddl/placement/bundle.go

```diff
@@ -49,6 +50,200 @@ func NewBundle(id int64) *Bundle {
+	}
+
+	if len(options.LeaderConstraints) > 0 || len(options.LearnerConstraints) > 0 || len(options.FollowerConstraints) > 0 || len(options.VoterConstraints) > 0 || options.Learners > 0 || options.Voters > 0 {
+		return nil, fmt.Errorf("%w: should be PRIMARY_REGION=.. REGIONS=.. FOLLOWERS=.. SCHEDULE=.., mixed other constraints into options %s", ErrInvalidPlacementOptions, options)
+	}
+
+	primaryRegion := strings.TrimSpace(options.PrimaryRegion)
+	if len(primaryRegion) == 0 {
+		return nil, fmt.Errorf("%w: empty primary region", ErrInvalidPlacementOptions)
+	}
+
+	var regions []string
+	if k := strings.TrimSpace(options.Regions); len(k) > 0 {
+		regions = strings.Split(k, ",")
+		for i, r := range regions {
+			regions[i] = strings.TrimSpace(r)
+		}
+	}
+
+	followers := options.Followers
+	if followers == 0 {
+		followers = 2
+	}
+	schedule := options.Schedule
+
+	var constraints Constraints
+	var err error
+
+	Rules := []*Rule{}
+	switch strings.ToLower(schedule) {
+	case "", "even":
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid PrimaryRegion '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+	case "majority_in_primary":
+		// number of quorum members, without the leader
+		primary := (followers + 1) / 2
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid PrimaryRegion, '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+		Rules = append(Rules, NewRule(Follower, primary, constraints))
+		// even split the remaining followers
+		followers = followers - primary
+	default:
+		return nil, fmt.Errorf("%w: unsupported schedule %s", ErrInvalidPlacementOptions, schedule)
+	}
+
+	if uint64(len(regions)) > followers {
+		return nil, fmt.Errorf("%w: remain %d region to schedule, only %d follower left", ErrInvalidPlacementOptions, uint64(len(regions)), followers)
+	}
+
+	if len(regions) == 0 {
+		constraints, err := NewConstraints(nil)
+		if err != nil {
+			return nil, err
+		}
+		Rules = append(Rules, NewRule(Follower, followers, constraints))
+	} else {
+		count := followers / uint64(len(regions))
+		rem := followers - count*uint64(len(regions))
+		for _, region := range regions {
+			constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", region)})
+			if err != nil {
+				return nil, fmt.Errorf("%w: invalid region of 'Regions', '%s'", err, region)
+			}
+			replica := count
+			if rem > 0 {
+				replica += 1
+				rem--
+			}
+			Rules = append(Rules, NewRule(Follower, replica, constraints))
+		}
+	}
+
+	return &Bundle{Rules: Rules}, nil
+}
+
+// NewBundleFromOptions will transform options into the bundle.
+func NewBundleFromOptions(options *model.PlacementSettings) (*Bundle, error) {
+	var isSyntaxSugar bool
+
+	if options == nil {
```

Comment:
nil checker is duplicate with the one in `NewBundleFromSugarOptions`

Reply:
It is exported, i.e. it may serve as a standalone method. So the check is intended.